### ISSUE TITLE
DuckDB example C++ program crashes for persistent database

### ISSUE THREAD

**Issue by sroycode** [cwida/duckdb#568]
I was checking out duckDB using the simple example program (on ubuntu 18),

```
#include "duckdb.hpp"
  
using namespace duckdb;

int main() {
  DuckDB db("./data/pdata");
  Connection con(db);
  con.Query("CREATE TABLE IF NOT EXISTS pdata (record_id BIGINT PRIMARY KEY , price DOUBLE, shop VARCHAR, category VARCHAR, name VARCHAR, brand VARCHAR, color VARCHAR, model VARCHAR);");
  auto result = con.Query("COPY pdata(record_id,price,shop,category,name,brand,color,model) FROM 'items.txt' ( DELIMITER '|' );");
  result->Print();
}
```
I tried to load a few csv files , each with about 10,000 records. After the fourth run, it crashes saying

```
terminate called after throwing an instance of 'duckdb::Exception'
  what():  Column length mismatch in table load!
Command terminated by signal 6
```
Incidentally , if I load a million records at a time, it still crashes after 5-6 iterations. 
Memory usage is not high.
What am I doing wrong please ?

**Comment by Mytherin** [comment]
This seems like a bug in the checkpointing process of the database. That is an internal system error. Would it be possible for you to share the CSV files?

**Comment by sroycode** [comment]
Generated randomly using a shellscript , `loadme` is the name of the program
I have uploaded a sample items.txt file and the generating shellscript here

https://gist.github.com/sroycode/74f0b5da73f73fca952aa45fa7ef0776

**Comment by Mytherin** [comment]
Thanks! I will have a look.

**Comment by Mytherin** [comment]
Solved this issue in ea2bb44236c57e2b7e53624af99c9c0f2c72bd07, but found another bug in string checkpointing afterwards. I will take a look at that tomorrow. This was a nice test case, thanks :)

**Comment by sroycode** [comment]
Hi:
Thanks, will wait. 
By the way, I am using upload batch size of around 1000 records , though there are atomic updates too much lesser in number. Is this optimal for say a total of 500M records ?

**Comment by Mytherin** [comment]
Fixed the problem in 6f45ee581a424f264dcea9341cc029d45746cf54. It was related to the free list used in the checkpoint. The example program should now work. I have integrated the example program in a test [here](https://github.com/cwida/duckdb/blob/6f45ee581a424f264dcea9341cc029d45746cf54/test/sql/storage/test_repeated_checkpoint.cpp) and have run it until counter=1000 without problems. 

@sroycode 
I would say 1000 records is a small batch size, especially if you want to have a disk-backed database. For optimal performance larger batches are better. Perhaps 100K records or so.

I should also mention that the way the example program operates is not a good fit for the current implementation of DuckDB. Restarting the database is quite expensive now as we don't have incremental checkpointing yet. Restarting the database can trigger a full checkpoint of the database, which involves a lot of Disk IO.
